awx/main/tasks.py

```diff
@@ -901,235 +953,253 @@ def get_instance_timeout(self, instance):
-                dispatcher.dispatch(event_data)
+    def post_run_hook(self, instance, status):
+        '''
+        Hook for any steps to run before job/task is marked as complete.
+        '''
 
-            return OutputEventFilter(event_callback)
-        else:
-            def event_callback(event_data):
-                event_data.setdefault(self.event_data_key, instance.id)
-                dispatcher.dispatch(event_data)
+    def final_run_hook(self, instance, status, private_data_dir, fact_modification_times):
+        '''
+        Hook for any steps to run after job/task is marked as complete.
+        '''
 
-            return OutputVerboseFilter(event_callback)
+    def event_handler(self, event_data):
+        '''
+        Ansible runner callback for events
+        '''
+        should_write_event = False
+        dispatcher = CallbackQueueDispatcher()
+        event_data.setdefault(self.event_data_key, self.instance.id)
+        dispatcher.dispatch(event_data)
+        self.event_ct += 1
 
-    def pre_run_hook(self, instance, **kwargs):
         '''
-        Hook for any steps to run before the job/task starts
+        Handle artifacts
         '''
+        if event_data.get('event_data', {}).get('artifact_data', {}):
+            self.instance.artifacts = event_data['event_data']['artifact_data']
+            self.instance.save(update_fields=['artifacts'])
+
+        return should_write_event
 
-    def post_run_hook(self, instance, status, **kwargs):
+    def cancel_callback(self):
         '''
-        Hook for any steps to run before job/task is marked as complete.
+        Ansible runner callback to tell the job when/if it is canceled
         '''
+        self.instance = self.update_model(self.instance.pk)
+        if self.instance.cancel_flag or self.instance.status == 'canceled':
+            cancel_wait = (now() - self.instance.modified).seconds if self.instance.modified else 0
+            if cancel_wait > 5:
+                logger.warn('Request to cancel {} took {} seconds to complete.'.format(self.instance.log_format, cancel_wait))
+            return True
+        return False
 
-    def final_run_hook(self, instance, status, **kwargs):
+    def finished_callback(self, runner_obj):
         '''
-        Hook for any steps to run after job/task is marked as complete.
+        Ansible runner callback triggered on finished run
         '''
+        dispatcher = CallbackQueueDispatcher()
+        event_data = {
+            'event': 'EOF',
+            'final_counter': self.event_ct,
+        }
+        event_data.setdefault(self.event_data_key, self.instance.id)
+        dispatcher.dispatch(event_data)
+
+    def status_handler(self, status_data, runner_config):
+        '''
+        Ansible runner callback triggered on status transition
+        '''
+        if status_data['status'] == 'starting':
+            self.instance = self.update_model(self.instance.pk, job_args=json.dumps(runner_config.command),
```

Comment:
I know that safe_args wasn't needed due to the extra_vars anymore, are other forms of unsafe data handled by runner?

Reply:
Not that I'm aware of. Wanted to see if anyone else could think of any.